README

                      Template Toolkit LaTeX Plugin

                            Version 3.00_05

                           13 October 2007

        Copyright (C) 2006-2007 Andrew Ford.   All Rights Reserved
        Copyright (C) 1996-2007 Andy Wardley.  All Rights Reserved

          This is free software; you can redistribute it and/or
             modify it under the same terms as Perl itself.


DESCRIPTION
-----------

The Template Toolkit is a collection of modules which implement a
fast, flexible, powerful and extensible template processing system.
The LaTeX plugin provides the facility to filter content through
LaTeX, provided of course that the content is valid LaTeX source.  

The LaTeX plugin takes care of running and re-running latex so that
forward references, tables of contents, and lists of figures and
tables are resolved.  It will also run bibtex and makeindex if it
detects that a bibliography or in index have been specified, and will
re-run latex again one or more times until the formatting of the
document has stabilized.  (This functionality is new in version 3.00
of the module.)



QUICK START
-----------

The latest version of the Template Toolkit LaTeX Plugin can be
retrieved from:

    http://www.cpan.org/modules/by-module/Template-Latex/

Fetch and install Template Toolkit (2.16 or newer) if you don't
already have it installed.  Available from CPAN in:

    http://www.cpan.org/authors/Andy_Wardley/

To install the Template Toolkit Latex Plugin:

    tar zxf Template-Latex-3.00_05.tar.gz
    cd Template-Latex-3.00_05
    perl Makefile.PL
    make
    make test
    make install    

The Makefile.PL will prompt for additional configuration options,
including the installation of optional template libraries, HTML
documentation and examples.  You can safely answer 'n' to all of these
questions for a quick and basic installation.  

If you're running ActivePerl on a Win32 platform then you can use the
Perl Package Manager (PPM) to install the Template Toolkit.  Chris
Winters maintains a repository of pre-compiled PPM packages which contains
the Template Toolkit, AppConfig and others.  For further information, see:

    http://openinteract.sourceforge.net/


For further details on installation, see the separate INSTALL file.


WHAT'S NEW?
-----------

Version 3.00 is a complete rewrite of Template-Latex.

Version 2.16 was the first release of Template-Latex.  It was created
by moving the code out of the core Template Toolkit distribution
(version 2.15).

See the Changes file for further details of the changes in these
releases.


GENERAL FEATURES
----------------

Some of the key features of the Template Toolkit Latex Plugin are listed below.
See the documentation for further detail.


DOCUMENTATION
-------------

The Template Toolkit is provided with enough documentation to keep all
but the most voracious reader happy for quite some time.

The 'Changes' file in the distribution directory documents all visible
changes between versions of the Template Toolkit.  See the section 
'VERSION COMPATABILITY' below for further details.

The 'TODO' file, also in the distribution directory, lists known bugs,
planned enhancements and possible new features for future versions.

The rest of the documentation is distributed in Pod and HTML formats.
The Pod pages are installed when you 'make install' and can be viewed
using 'perldoc', e.g.


SUPPORT
-------

The Template Toolkit mailing list provides a forum for discussing
issues relating to the use and abuse of the Template Toolkit.  There
are a number of knowledgeable and helpful individuals who frequent the
list (including the author) who can often offer help or suggestions.
Please respect their time and patience by checking the documentation
and/or mailing list archives before asking questions that may already
have been answered.   

To subscribe to the mailing list, send an email to:

    templates-request@template-toolkit.org

with the message 'subscribe' in the body.  You can also use the web 
interface to subscribe or browse the archives:

    http://www.template-toolkit.org/mailman/listinfo/templates

A low-volume, moderated mailing list exists for announcements about new
releases of the Template Toolkit and related products.  To subscribe, send
an email to:

    templates-announce-request@template-toolkit.org

with the message 'subscribe' in the body.  A web interface also exists
for subscription and browsing the archives:

    http://www.template-toolkit.org/mailman/listinfo/templates-announce

For information about commercial support and consultancy for the Template 
Toolkit, please contact the author.


AUTHOR
------

The Template Toolkit Latex plugin was originally written as part of
the Template Toolkit distribution by Andy Wardley <abw@wardley.org>
with the assistance and contributions from many other people. In 2006
the Latex plugin was separated out into its own distribution and
Andrew Ford <a.ford@ford-mason.co.uk> the took over as maintainer.

See Template::Manual::Credits for details.


COPYRIGHT
---------

Copyright (C) 2006-2007 Andrew Ford.   All Rights Reserved.
Copyright (C) 1996-2007 Andy Wardley.  All Rights Reserved.

This is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.



Changes
#================================================= -*- indented-text -*-
#
# Changes 
#
# DESCRIPTION
#   Revision history for the Template Toolkit Latex plugin.
#
# AUTHOR
#   Andrew Ford   <A.Ford@ford-mason.co.uk>
#
#
#------------------------------------------------------------------------
# $Id: Changes 86 2009-03-11 17:45:23Z ford $
#========================================================================

#------------------------------------------------------------------------
# Version 3.02 - 2009-03-12
#------------------------------------------------------------------------

* fixed skip_all call (use Template::Test rather than Test::More - 
  despite Template::Test being incompatible with TAP)

#------------------------------------------------------------------------
# Version 3.01 - 2009-03-11
#------------------------------------------------------------------------

* tidied up tests to avoid spurious failures (skip_all if the 'dvitype'
  program is not available)

* expanded documentation

#------------------------------------------------------------------------
# Version 3.00 - 2009-03-09
#------------------------------------------------------------------------

* first non-developer release - no changes from 3.00_05


#------------------------------------------------------------------------
# Version 3.00_05 - 2007-10-13
#------------------------------------------------------------------------

* Removed "use Template" from Makefile.PL as that was unnecessary and
  was causing build failures where Template was not already installed.


#------------------------------------------------------------------------
# Version 3.00_04 - 2007-10-13
#------------------------------------------------------------------------

* Added a test script for the "table" method

* Updated dependency on LaTeX::Driver to 0.07 as 0.06 failed to build
  on cpantesters machines

* mark 11-plugin-errors.t as all skipped, as the error messages that
  are checked for are not what the driver module produces - I need to
  review this test script and change some of the individual checks to
  do meaningful things

#------------------------------------------------------------------------
# Version 3.00_03 - 2007-10-03
#------------------------------------------------------------------------

* Added a "table" method to the plugin to interface to the
  LaTeX::Table module.  This will generate the LaTeX commands to
  format a table from raw data.

#------------------------------------------------------------------------
# Version 3.00_02 - 2007-10-03
#------------------------------------------------------------------------

* Restructured again so that Template::Latex calls the
  Template::Plugin::Latex module, which in turn builds on
  LaTeX::Driver to format the document.  (Removed the
  Template::Latex::Paths module as the paths are now held in the
  LaTeX::Driver module.)

* Expanded the test suite and the documentaton.

#------------------------------------------------------------------------
# Version 3.00_01 - 2007-09-13
#------------------------------------------------------------------------

* I have restructured the plugin fairly extensively, and added
  radically new features, so this release has a new major version
  number, i.e. 3.x.

* Restructured Template::Latex to run latex multiple times if
  necessary and also to run bibtex if citations are detected and
  makeindex if indexing is detected.

* The TEXINPUTS environment variable is set up to include the original
  directory so that include files can be found there.

* Change the format of the commands passed to system() so that the cd
  into the temporary directory is done in the command rather than in
  the module.  That way the module does not need to change back to the
  original directory, simplifying error handling.  
  (The Win32 idiom is: 'cmd /c "cd DIR && COMMAND"')

* Split the driver code in Template::Latex out into its own module
  (Template::Latex::Driver).

* The LaTeX macro \TTLATEX is defined when latex or pdflatex is run

* The static filter detex is defined by the plugin

#------------------------------------------------------------------------
# Version 2.17 - 2006-06-07
#------------------------------------------------------------------------

* Changes dependency on Template to version 2.15 as version 2.16 of
  Template, which was specified as the dependency has not been relased.

* Added files: Changes (this file) and TODO


#------------------------------------------------------------------------
# Version 2.16 - 2006-05-31
#------------------------------------------------------------------------

First release of the plugin as a separate distribution to the core
Template Toolkit distribution.  New maintainer: Andrew Ford.


lib/Template/Latex.pm
#============================================================= -*-perl-*-
#
# Template::Latex
#
# DESCRIPTION
#   Provides an interface to Latex from the Template Toolkit.
#
# AUTHOR
#   Andrew Ford    <a.ford@ford-mason.co.uk>  (current maintainer)
#   Andy Wardley   <abw@wardley.org>
#
# COPYRIGHT
#   Copyright (C) 2006-2007 Andrew Ford.   All Rights Reserved.
#   Copyright (C) 1996-2006 Andy Wardley.  All Rights Reserved.
#
#   This module is free software; you can redistribute it and/or
#   modify it under the same terms as Perl itself.
#
# HISTORY
#   * Latex plugin originally written by Craig Barratt, Apr 28 2001.
#   * Win32 additions by Richard Tietjen.
#   * Extracted into a separate Template::Latex module by Andy Wardley,
#     May 2006
#   * Removed the functionality to specify program pathnames on the FILTER call
#     Andrew Ford, 05 June 2006
#   * Now runs BibTeX and Makeindex if required
#     Andrew Ford, 05 June 2006
#   * Rewritten to set up TEXINPUTS so that inclusions can be found and to run
#     latex, bibtex and makeindex repeatedly as needed until references stabilize
#     (NOT YET FINISHED)
#     Andrew Ford, September 2007
#
#========================================================================
 
package Template::Latex;

use strict;
use warnings;
use base 'Template';
use Template::Exception;
use Template::Plugin::Latex;
use LaTeX::Driver;

our $VERSION  = 3.00_01;
our $DEBUG    = 0 unless defined $DEBUG;
our $ERROR    = '';
our $FILTER   = 'latex';        # default filter name
our $FORMAT   = '';             # output format (auto-detect if unset)
our @PROGRAMS = qw( latex pdflatex bibtex makeindex dvips ps2pdf xelatex );


sub new {
    my $class  = shift;
    my $config = @_ && ref $_[0] eq 'HASH' ? shift : { @_ };
    my $self   = $class->SUPER::new($config) || return;

#     # set default format from config option
#     $self->latex_format($config->{ LATEX_FORMAT })
#         if $config->{ LATEX_FORMAT };

#     # set latex paths from config options
#     $self->latex_paths({
#         latex     => $config->{ LATEX_PATH     },
#         pdflatex  => $config->{ PDFLATEX_PATH  },
#         dvips     => $config->{ DVIPS_PATH     },
#         ps2pdf    => $config->{ PS2PDF_PATH    },
#         bibtex    => $config->{ BIBTEX_PATH    },
#         makeindex => $config->{ MAKEINDEX_PATH },
#     });


    # install the latex filter
    Template::Plugin::Latex->new($self->context, {});

   #$self->define_filter( $self->context() );

    return $self;
}


#------------------------------------------------------------------------
# latex_format()
# latex_path()
# pdflatex_path()
# dvips_path()
#
# Methods to get/set the $FORMAT, $LATEX, $PDFLATEX and $DVIPS package
# variables that specify the default output format and the paths to
# the latex, pdflatex and dvips programs.
#------------------------------------------------------------------------

sub latex_format {
    my $class = shift;
    return @_ ? ($FORMAT = shift) : $FORMAT;
}

sub latex_path {
    my $class = shift;
    return LaTeX::Driver->program_path('latex', @_);
}

sub pdflatex_path {
    my $class = shift;
    return LaTeX::Driver->program_path('pdflatex', @_);
}

sub dvips_path {
    my $class = shift;
    return LaTeX::Driver->program_path('dvips', @_);
}

sub ps2pdf_path {
    my $class = shift;
    return LaTeX::Driver->program_path('ps2pdf', @_);
}

sub bibtex_path {
    my $class = shift;
    return LaTeX::Driver->program_path('bibtex', @_);
}

sub makeindex_path {
    my $class = shift;
    return LaTeX::Driver->program_path('makeindex', @_);
}

sub xelatex_path {
    my $class = shift;
    return LaTeX::Driver->program_path('xelatex', @_);
}

#------------------------------------------------------------------------
# latex_paths()
#
# Method to get/set the above all in one go.
#------------------------------------------------------------------------

sub latex_paths {
    my $class = shift;
    if (@_) {
        my $args = ref $_[0] eq 'HASH' ? shift : { @_ };

        LaTeX::Driver->program_path($_, $args->{$_})
            for qw( latex pdflatex bibtex makeindex dvips ps2pdf xelatex );
    }
    else {
        return {
            map { ( $_ => LaTeX::Driver->program_path($_) ) } @PROGRAMS
        };
    }
}

1;

__END__


#------------------------------------------------------------------------
# define_filter($context, $config)
#
# This method defines the latex filter in the context specified as the 
# first argument.  A list or hash ref of named parameters can follow
# providing default configuration options.  
#------------------------------------------------------------------------

# sub define_filter {
#     my $class   = shift;
#     my $context = shift;
#     my $default = @_ && ref $_[0] eq 'HASH' ? shift : { @_ };
#     my $filter  = $default->{ filter } || $FILTER;

#     # default any config item not set to values in package variables
#     $default->{ format    } ||= $FORMAT;
#     $default->{ latex     } ||= $LATEX;
#     $default->{ pdflatex  } ||= $PDFLATEX;
#     $default->{ dvips     } ||= $DVIPS;
#     $default->{ ps2pdf    } ||= $PS2PDF;
#     $default->{ bibtex    } ||= $BIBTEX;
#     $default->{ makeindex } ||= $MAKEINDEX;

#     # define a factory subroutine to be called when the filter is used.
#     my $factory = sub { 
#         my $context = shift;
#         my $config  = @_ && ref $_[0] eq 'HASH' ? pop : { };

#         # merge any configuration parameters specified when the filter
#         # is used with the defaults provided when the filter was defined
#         $config->{ $_ } ||= $default->{ $_ } 
#             for (qw( format latex pdflatex dvips ps2pdf bibtex makeindex ));

#         # output file can be specified as the first argument
#         $config->{ output } = shift if @_;

#         $config->{ DEBUG } ||= $DEBUG;

#         # return an anonymous filter subroutine which calls the real
#         # filter() method passing the context and merged config params
#         return sub {
#             Template::Latex::Driver->run(shift, $context, $config);
#         };
#     };

#     # install the filter factory in the context
#     $context->define_filter( $filter => $factory, 1 );
#     $context->define_filter( detex => \&detex_filter );
# }


1;

__END__


=head1 NAME

Template::Latex - Latex support for the Template Toolkit

=head1 SYNOPSIS

    use Template::Latex;
    
    my $tt = Template::Latex->new({
        INCLUDE_PATH  => '/path/to/templates',
        OUTPUT_PATH   => '/path/to/pdf/output',
        LATEX_FORMAT  => 'pdf',
    });
    my $vars = {
        title => 'Hello World',
    }
    $tt->process('example.tt2', $vars, 'example.pdf', binmode => 1)
        || die $tt->error();

=head1 DESCRIPTION

The Template::Latex module is a wrapper of convenience around the
Template module, providing additional support for generating PDF,
PostScript and DVI documents from LaTeX templates.

You use the Template::Latex module exactly as you would the Template
module.  

    my $tt = Template::Latex->new(\%config);
    $tt->process($input, \%vars, $output)
        || die $t->error();

It supports a number of additional configuration parameters. The
C<LATEX_PATH>, C<PDFLATEX_PATH> and C<DVIPS_PATH> options can be used
to specify the paths to the F<latex>, F<pdflatex> and F<dvips> program
on your system, respectively.  These are usually hard-coded in the
Template::Latex C<$LATEX>, C<$PDFLATEX> and C<$DVIPS> package
variables based on the values set when you run C<perl Makefile.PL> to
configure Template::Latex at installation time.  You only need to
specify these paths if they've moved since you installed
Template::Latex or if you want to use different versions for some
reason.

    my $tt = Template::Latex->new({
        LATEX_PATH    => '/usr/bin/latex',
        PDFLATEX_PATH => '/usr/bin/pdflatex',
        DVIPS_PATH    => '/usr/bin/dvips',
    });

It also provides the C<LATEX_FORMAT> option to specify the default
output format.  This can be set to C<pdf>, C<ps> or C<dvi>.

    my $tt = Template::Latex->new({
        LATEX_FORMAT  => 'pdf',
    });

The C<latex> filter is automatically defined when you use the
Template::Latex module.  There's no need to load the Latex plugin in
this case, although you can if you want (e.g. to set some
configuration defaults).  If you're using the regular Template module
then you should first load the Latex plugin to define the C<latex>
filter.

    [% USE Latex %]
    [% FILTER latex('example.pdf') %]
    ...LaTeX doc...
    [% END %]

=head1 PUBLIC METHODS

The Template::Latex module is a subclass of the Template module and
inherits all its methods.  Please consult the documentation for the
L<Template> module for further information on using it for template
processing.  Wherever you see C<Template> substitute it for
C<Template::Latex>.

In addition to those inherted from the Template module, the following
methods are also defined.

=head2 latex_paths()

Method to get or set the paths to the F<latex>, F<pdflatex> and
F<dvips> programs.  These values are stored in the Template::Latex
C<$LATEX>, C<$PDFLATEX> and C<$DVIPS> package variables, respectively.
It can be called as either a class or object method.

    Template::Latex->latex_paths({
        latex    => '/usr/bin/latex',
        pdflatex => '/usr/bin/pdflatex',
        dvips    => '/usr/bin/dvips',
    });

    my $paths = Template::Latex->latex_paths();
    print $paths->{ latex };    # /usr/bin/latex

=head2 latex_path()

Method to get or set the C<$Template::Latex::LATEX> package
variable which defines the location of the F<latex> program on your
system.  It can be called as a class or object method.

    Template::Latex->latex_path('/usr/bin/latex');
    print Template::Latex->latex_path();   # '/usr/bin/latex'

=head2 pdflatex_path()

Method to get or set the C<$Template::Latex::PDFLATEX> package
variable which defines the location of the F<pdflatex> program on your
system.  It can be called as a class or object method.

    Template::Latex->pdflatex_path('/usr/bin/pdflatex');
    print Template::Latex->pdflatex_path();   # '/usr/bin/pdflatex'

=head2 dvips_path()

Method to get or set the C<$Template::Latex::DVIPS> package
variable which defines the location of the F<dvips> program on your
system.  It can be called as a class or object method.

    Template::Latex->dvips_path('/usr/bin/dvips');
    print Template::Latex->dvips_path();   # '/usr/bin/dvips'

=head2 bibtex_path()

Method to get or set the C<$Template::Latex::BIBTEX> package
variable which defines the location of the F<bibtex> program on your
system.  It can be called as a class or object method.

    Template::Latex->bibtex_path('/usr/bin/bibtex');
    print Template::Latex->bibtex_path();   # '/usr/bin/bibtex'

=head2 makeindex_path()

Method to get or set the C<$Template::Latex::MAKEINDEX> package
variable which defines the location of the F<makeindex> program on your
system.  It can be called as a class or object method.

    Template::Latex->makeindex_path('/usr/bin/makeindex');
    print Template::Latex->makeindex_path();   # '/usr/bin/makeindex'

=head1 INTERNALS

This section is aimed at a technical audience.  It documents the
internal methods and subroutines as a reference for the module's
developers, maintainers and anyone interesting in understanding how it
works.  You don't need to know anything about them to use the module
and can safely skip this section.

=head2 define_filter($context,\%config)

This class method installs the C<latex> filter in the context passed
as the first argument.  The second argument is a hash reference
containing any default filter parameters (e.g. those specified when
the Template::Plugin::Latex plugin is loaded via a C<USE> directive).

    Template::Latex->define_filter($context, { format => 'pdf' });

The filter is installed as a I<dynamic filter factory>.  This is just
a fancy way of saying that the filter generates a new filter
subroutine each time it is used to account for different invocation
parameters.  The filter subroutine it creates is effectively a wrapper
(a "closure" in technical terms) around the C<filter()> subroutine
(see below) which does the real work.  The closure keeps track of any
configuration parameters specified when the filter is first defined
and/or when the filter is invoked.  It passes the merged configuration
as the second argument to the C<filter()> subroutine (see below).

See the L<Template::Filters> module for further information on how
filters work.

=head2 filter($text,\%config)

This is the main LaTeX filter subroutine which is called by the
Template Toolkit to generate a LaTeX document from the text passed as
the first argument.  The second argument is a reference to a hash
array of configuration parameters.  These are usually provided by the
filter subroutine that is generated by the filter factory.

    Template::Latex::filter($latex, { 
        latex    => '/usr/bin/latex',
        pdflatex => '/usr/bin/pdflatex',
        dvips    => '/usr/bin/dvips',
        output   => 'example.pdf',
    });

=head2 throw($message)

Subroutine which throws a L<Template::Exception> error using C<die>.
The exception type is set to C<latex>.

    Template::Latex::throw("I'm sorry Dave, I can't do that");

=head2 debug($message)

Debugging subroutine which print all argument to STDERR.  Set the 
C<$DEBUG> package variable to enable debugging messages.

    $Template::Latex::DEBUG = 1;

=head1 AUTHOR

Andrew Ford E<lt>a.ford@ford-mason.co.ukE<gt> (current maintainer)

Andy Wardley E<lt>abw@wardley.orgE<gt> L<http://wardley.org/>

The original Latex plugin on which this is based was written by Craig
Barratt with additions for Win32 by Richard Tietjen.

=head1 COPYRIGHT

Copyright (C) 1996-2006 Andy Wardley.  All Rights Reserved.

Copyright (C) 2006-2007 Andrew Ford.  All Rights Reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=head1 SEE ALSO

L<Template::Plugin::Latex>

=cut

# Local Variables:
# mode: perl
# perl-indent-level: 4
# indent-tabs-mode: nil
# End:
#
# vim: expandtab shiftwidth=4:


Makefile.PL
#!/usr/bin/perl -w                                         # -*- perl -*-

use strict;

use lib qw( ./lib );
use Config;
use File::Spec::Functions qw( catfile );
use ExtUtils::MakeMaker;
use Cwd;

select STDERR;
$| = 1;
select STDOUT;

our $QUIET  = 0;
our $ACCEPT = 0;
our $WIN32  = ($^O eq 'MSWin32');
our $MODVERSION = get_module_version();

#------------------------------------------------------------------------

message(<<EOF);

Template-Plugin-Latex v$MODVERSION
------------------------------

The Template-Latex modules add LaTeX support for the Template Toolkit.

They allow you to create LaTeX documents using the full power of the
Template Toolkit.  The generated output can then be processed with
either 'pdflatex' or 'latex'.  If unresolved cross references,
bibliographic references or index definitions are found then 'bibtex'
or 'makeindex' will be run as appropriate and 'latex' or 'pdflatex'
re-run as necessary.  The output will be postprocessed with the
'dvips' and 'ps2pdf' programs if necessary to create PDF, DVI or
PostScript documents.

To use the Template-Latex module you will first need to install LaTeX on
your system and make sure the above programs are available.  

EOF

#------------------------------------------------------------------------

if (! find_program($ENV{PATH}, "latex") and ! find_program($ENV{PATH}, "pdflatex")) {
    my $continue = ttprompt('You don\'t seem to have LaTeX installed.  Continue anyway?', 'N');
    die("Aborting as LaTeX is not installed\n") unless uc($continue) =~ /^Y/;
}


#------------------------------------------------------------------------

my %opts = (
    'NAME'	       => 'Template-Plugin-Latex',
    'VERSION_FROM' => 'lib/Template/Plugin/Latex.pm',
    'PMLIBDIRS'    => [ 'lib' ], 
    'PREREQ_PM'    => { 
        'Template'      => 2.16,
	'LaTeX::Driver' => 0.07,
	'LaTeX::Encode' => 0.02,
	'LaTeX::Table'  => 0,
    },
    'dist'         => {
        'COMPRESS' => 'gzip',
        'SUFFIX'   => 'gz',
    },
    'clean'        => {
        'FILES'    => join(' ', qw( t/output/test1.pdf 
                                    t/output/test1.ps 
                                    t/output/test1.dvi
                                    t/output/test2 
                                    t/output/test2.pdf 
                                    t/output/test2.ps 
                                    t/output/test2.dvi )),
    },
);

if ($ExtUtils::MakeMaker::VERSION >= 5.43) {
    $opts{ AUTHOR   } = 'Andrew Ford <a.ford@ford-mason.co.uk>';
    $opts{ ABSTRACT } = 'Latex support for the Template Toolkit',
}

WriteMakefile(%opts);



#------------------------------------------------------------------------
# find_program($path, $prog)
#
# Find a program, $prog, by traversing the given directory path, $path.
# Returns full path if the program is found.
#
# Written by Craig Barratt, Richard Tietjen add fixes for Win32.
#
# abw changed name from studly caps findProgram() to find_program() :-)
#------------------------------------------------------------------------

sub find_program {
    my($path, $prog) = @_;

    foreach my $dir ( split($Config{path_sep}, $path) ) {
        my $file = File::Spec->catfile($dir, $prog);
        if ( !$WIN32 ) {
            return $file if ( -x $file );
        } else {
            # Windows executables end in .xxx, exe precedes .bat and .cmd
            foreach my $dx ( qw/exe bat cmd/ ) {
                return "$file.$dx" if ( -x "$file.$dx" );
            }
        }
    }
}


#------------------------------------------------------------------------
# message($text)
#
# Print message unless quiet mode.
#------------------------------------------------------------------------

sub message {
    return if $QUIET;
    print @_;
}


#------------------------------------------------------------------------
# ttprompt($message, $default)
#------------------------------------------------------------------------

sub ttprompt {
    my ($msg, $def)=@_;
    my $ISA_TTY = -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT)) ; # Pipe?
    my $dispdef = defined $def ? "[$def] " : " ";
    $def = defined $def ? $def : "";
    my $ans = '';
    local $|=1;
    print "$msg $dispdef" unless $QUIET;
    if ($ACCEPT || ! $ISA_TTY) {
        print "$def\n" unless $QUIET;
    }
    else {
        chomp($ans = <STDIN>);
    }
    return ($ans ne '') ? $ans : $def;
}


#------------------------------------------------------------------------
# get_module_version()
#
# Reads the module version from the Template::Plugin::Latex module file
#------------------------------------------------------------------------

sub get_module_version {
    if (open(MODULE, catfile('lib','Template','Plugin','Latex.pm'))) {
	while (<MODULE>) {
	    return $1 if /VERSION = "?(\d+\.[\d_]+)/;
	}
    }
    return "<unkown version>";
}
